napari/utils/key_bindings/dispatch.py

```diff
@@ -0,0 +1,353 @@
+
+from napari.utils.key_bindings.constants import (
+    BASE_KEY_MASK,
+    VALID_KEYS,
+    DispatchFlags,
+)
+from napari.utils.key_bindings.register import (
+    KeyBindingEntry,
+    NapariKeyBindingsRegistry,
+)
+from napari.utils.key_bindings.util import create_conflict_filter, key2mod
+
+logger = logging.getLogger(__name__)
+
+
+def next_active_match(
+    entries: List[KeyBindingEntry], context: Mapping[str, object]
+) -> Optional[KeyBindingEntry]:
+    """Find and yield active matches while traversing through the entries.
+
+    See `NAP 7 <https://napari.org/dev/naps/7-key-binding-dispatch.html#key-binding-properties>`_.
+    for more information.
+
+    Parameters
+    ----------
+    entries: List[KeyBindingEntry]
+        Pre-sorted entries to search through.
+    context: Mapping[str, object]
+        Context with which to evaluate the entries.
+
+    Yields
+    ------
+    match: KeyBindingEntry
+        Next match encountered.
+    """
+    ignored_commands = []
+
+    for entry in reversed(entries):
+        if entry.when is None or entry.when.eval(context):
+            if entry.block_rule:
+                break
+            elif entry.negate_rule:
+                command_id = entry.command_id[1:]
+                ignored_commands.append(command_id)
+            elif entry.command_id not in ignored_commands:
+                yield entry
+
+
+def find_active_match(
+    entries: List[KeyBindingEntry], context: Mapping[str, object]
+) -> Optional[KeyBindingEntry]:
+    """Find and return the first active match.
+
+    See `NAP 7 <https://napari.org/dev/naps/7-key-binding-dispatch.html#key-binding-properties>`_.
+    for more information.
+
+    Parameters
+    ----------
+    entries: List[KeyBindingEntry]
+        Pre-sorted entries to search through.
+    context: Mapping[str, object]
+        Context with which to evaluate the entries.
+
+    Returns
+    -------
+    match: KeyBindingEntry or None
+        First match encountered, if found.
+    """
+    return next(next_active_match(entries, context), None)
+
+
+def has_conflicts(
+    key: int,
+    keymap: Dict[int, List[KeyBindingEntry]],
+    context: Mapping[str, object],
+) -> bool:
+    """Check if the given key has a conflict. Only works for the first part of a key binding.
+
+    See `NAP 7 <https://napari.org/dev/naps/7-key-binding-dispatch.html#key-binding-properties>`_.
+    for more information.
+
+    Parameters
+    ----------
+    key: int
+        Key, modifier, or key combo to check.
+    keymap: Dict[int, List[KeyBindingEntry]]
+        Keymap to check for conflicts.
+    context: Mapping[str, object]
+        Context with which to evaluate the entries.
+
+    Returns
+    -------
+    bool
+        If the given key has other active conflicts.
+    """
+    conflict_filter = create_conflict_filter(key)
+
+    for _, entries in filter(
+        lambda kv: conflict_filter(kv[0]), keymap.items()
+    ):
+        if find_active_match(entries, context):
```

Comment:
This will emit a deprecationWarning and fail in future Python version because.... (see other comment)